src/psyclone/psyGen.py

```diff
@@ -4250,6 +4266,188 @@ def nodes_to_code_block(parent, statements):
+                                        Fortran2003.Specification_Part)
+        except (ValueError, IndexError):
+            raise GenerationError("Unexpected kernel AST when generating "
+                                  "'{0}' kernel schedule. Could not find the "
+                                  "specification part.".format(name))
+
+        try:
+            subroutines = first_type_match(mod_content,
+                                           Fortran2003.Module_Subprogram_Part)
+            subroutine = search_subroutine(subroutines.content, name)
+        except (ValueError, IndexError):
+            raise GenerationError("Unexpected kernel AST. Could not find "
+                                  "subroutine: {0}".format(name))
+
+        try:
+            sub_spec = first_type_match(subroutine.content,
+                                        Fortran2003.Specification_Part)
+            decl_list = sub_spec.content
+            arg_list = subroutine.content[0].items[2].items
+        except ValueError:
+            # Subroutine without declarations, continue with empty lists.
+            decl_list = []
+            arg_list = []
+        except (IndexError, AttributeError):
+            # Subroutine without argument list, continue with empty list.
+            arg_list = []
+        finally:
+            self.process_declarations(new_schedule, decl_list, arg_list)
+
+        try:
+            sub_exec = first_type_match(subroutine.content,
+                                        Fortran2003.Execution_Part)
+        except ValueError:
+            pass
+        else:
+            self.process_nodes(new_schedule, sub_exec.content, sub_exec)
+
+        return new_schedule
+
+    def process_declarations(self, parent, nodes, arg_list):
+        '''
+        Transform the variable declarations in fparser2 parse tree into
+        symbols in the PSyIR parent node symbol table.
+
+        :param parent: PSyIR node in which to insert the symbols found.
+        :type parent: :py:class:`psyclone.psyGen.KernelSchedule`
+        :param nodes: fparser2 AST nodes to search for declaration statements.
+        :type nodes: list of :py:class:`fparser.two.utils.Base`
+        :param arg_list: fparser2 AST node containing the argument list.
+        :type arg_list: :py:class:`fparser.Fortran2003.Dummy_Arg_List`
+        :raises NotImplementedError: The provided declarations contain
+                                     attributes which are not supported yet.
+        '''
+        from fparser.two.utils import walk_ast
+        from fparser.two import Fortran2003
+        for decl in walk_ast(nodes, [Fortran2003.Type_Declaration_Stmt]):
+            # Parse data type, currently restricted to integers and reals.
+            datatype = 'unknown'
+            if str(decl.items[0]).lower().startswith('real'):
+                datatype = 'real'
+            elif str(decl.items[0]).lower().startswith('integer'):
+                datatype = 'integer'
```

Comment:
Please add `else NotImplementedError` so that we don't silently ignore things that we don't understand.

Reply:
Done